src/node_crypto.cc

```diff
@@ -685,26 +728,24 @@ void SecureContext::AddCACert(const FunctionCallbackInfo<Value>& args) {
     return env->ThrowTypeError("CA certificate argument is mandatory");
   }
 
-  if (!sc->ca_store_) {
-    sc->ca_store_ = X509_STORE_new();
-    newCAStore = true;
+  BIO* bio = LoadBIO(env, args[0]);
+  if (!bio) {
+    return;
   }
 
-  unsigned cert_count = 0;
-  if (BIO* bio = LoadBIO(env, args[0])) {
-    while (X509* x509 =
-        PEM_read_bio_X509(bio, nullptr, CryptoPemCallback, nullptr)) {
-      X509_STORE_add_cert(sc->ca_store_, x509);
-      SSL_CTX_add_client_CA(sc->ctx_, x509);
-      X509_free(x509);
-      cert_count += 1;
+  X509_STORE* cert_store = SSL_CTX_get_cert_store(sc->ctx_);
+  while (X509* x509 =
+             PEM_read_bio_X509(bio, nullptr, CryptoPemCallback, nullptr)) {
+    if (cert_store == root_cert_store) {
+      cert_store = NewRootCertStore();
```

Comment:
If the patch does change that behaviour then I agree that it's incorrect, but are you sure that it does?

The intent is that calling `AddRootCerts` followed by `AddCACert` should get you the root set plus the custom cert. Calling just `AddCACert` should get you just the custom cert. (Calling `AddCACert` followed by `AddRootsCerts` is confusing because the latter resets the store, but that's a different issue.)

These functions are wrapped by `lib/_tls_common.js` which calls `addRootCerts` only if `options.ca` is false in a boolean context. So I believe that setting 'ca' will still override the root certs, not augment them.

If I apply the following change and rerun the test with, and without, `ca:` commented then a connection to `www.google.com` fails when `ca` is set, as expected.

(Calling out to `www.google.com` in a test isn't good as you noted and I'd like to have a test for this behaviour, but for investigating whether things are currently broken it serves.)

```
diff --git a/test/parallel/test-tls-client-verify.js b/test/parallel/test-tls-client-verify.js
index 7844253..e1b5b44 100644
--- a/test/parallel/test-tls-client-verify.js
+++ b/test/parallel/test-tls-client-verify.js
@@ -124,8 +124,19 @@ function runTest(testIndex) {
   });
 }
 
-
-runTest(0);
+var caCert = fs.readFileSync(require('path').join(common.fixturesDir, 'keys', 'agent1-cert.pem'))
+console.error(caCert);
+
+var opts = {
+  host: "www.google.com",
+  port: 443,
+  //ca: caCert,
+  rejectUnauthorized: false,
+};
+var testClient = tls.connect(opts, function() {
+  console.error('connection to google.com authorized: ' + testClient.authorized);
+  runTest(0);
+})
```